cmd/server/main.go

```diff
@@ -17,31 +27,146 @@ func main() {
+	logger.G.Infof("Running server on %s", config.Params.Address)
+	server := http.Server{
+		Addr:    config.Params.Address,
+		Handler: getRouter(),
+	}
+	wg.Add(1)
+	go func() {
+		err = stopServer(&server, ctx, wg)
+	}()
+	err = server.ListenAndServe()
+	if err != nil && !errors.Is(err, http.ErrServerClosed) {
+		return err
+	}
+	return nil
+}
+
+func stopServer(server *http.Server, ctx context.Context, wg *sync.WaitGroup) error {
+	<-ctx.Done()
+	// Заставляем завершиться сервер и ждём его завершения
+	err := server.Shutdown(ctx)
+	if err != nil {
+		logger.G.Errorf("Failed to shutdown server: %v", err)
+	}
+	logger.G.Info("Server stop")
+	wg.Done()
+	return err
 }
 
 // getRouter конфигурация роутинга приложение
 func getRouter() chi.Router {
 	router := chi.NewRouter()
-	// Сохранение метрики
-	router.Post("/update/{type}/{name}/{value}", handlemetric.Handler)
+	// Устанавилваем мидлваре
+	router.Use(
+		cMiddleware.StripSlashes,          // Убираем лишние слеши
+		logger.LogRequests,                // Логируем данные запроса
+		middlewares.GZIPCompressResponse,  // Сжимаем ответ TODO исключить для роутов, которые будут возвращать не application/json или text/html. Проверять в мидлваре или компрессоре может быть не эффективно,так как заголовок с контентом может быть поставлен позже записи контента
+		logger.LogResponse,                // Логируем данные ответа
+		middlewares.GZIPDecompressRequest, // Разжимаем тело ответа
+	)
+	// Сохранение метрики по URL
+	router.Post("/update/{type}/{name}/{value}", handlemetric.URLHandler)
 	// Получение всех метрик
 	router.Get("/", getmetrics.Handler)
 	// Получение отдельной метрики
-	router.Get("/value/{type}/{name}", getmetric.Handler)
+	router.Get("/value/{type}/{name}", getmetric.URLHandler)
+
+	router.Group(func(r chi.Router) {
+		// Устанавилваем мидлваре с логированием запросов
+		r.Use(middlewares.JSONHeaders)
+		// Сохранение метрики с помощью JSON тела
+		r.Post("/update", handlemetric.JSONHandler)
+		// Получение отдельной метрики
+		r.Post("/value", getmetric.JSONHandler)
+	})
 	return router
 }
+
+type next func()
+
+// InitStore устанавливаем глобальное хранилище метрик.
+func InitStore(n next, ctx context.Context, wg *sync.WaitGroup) {
+	//Если указан путь к файлу, то будет создано хранилище с сохранением в файл, иначе будет создано хранилище в памяти
+	if config.Params.FileStorage != "" {
+		logger.G.Info("Set file store")
+		store, err := metrics.NewFileStorage(config.Params.FileStorage, config.Params.Restore, config.Params.StoreInterval == 0)
+		if err != nil {
+			logger.G.Fatal(err)
+		}
+		defer func() {
+			logger.G.Info("Close storage")
+			if dErr := store.FlushAndClose(); dErr != nil {
+				logger.G.Error(dErr)
+			}
+		}()
+		metrics.MeStore = store
+		ctx = context.WithValue(ctx, contextkeys.SyncInterval, config.Params.StoreInterval)
+		// Запускаем синхронизацию в файл
+		if !store.SyncMode {
+			wg.Add(1)
+			go func() {
+				store.Sync(ctx)
+				defer wg.Done()
```

Comment:
`defer` должен вызываться как можно раньше после входа в функцию. Если краш произойдет в `store.Sync(ctx)` defer ни когда не выполнится.

```suggestion
				defer wg.Done()
				store.Sync(ctx)
```

Reply:
Поправил

Reply:
Перенёс